sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java

```diff
@@ -93,13 +120,155 @@ public void populateDisplayData(DisplayData.Builder builder) {
     }
 
     /**
+     * Returns the number of shards that will be produced in the output.
+     *
+     * @see Write for more information
+     */
+    public int getNumShards() {
+      return numShards;
+    }
+
+    /**
      * Returns the {@link Sink} associated with this PTransform.
      */
     public Sink<T> getSink() {
       return sink;
     }
 
     /**
+     * Returns a new {@link Write.Bound} that will write to the current {@link Sink} using the
+     * specified number of shards.
+     *
+     * <p>This option should be used sparingly as it can hurt performance. See {@link Write} for
+     * more information.
+     *
+     * <p>A value less than or equal to 0 will be equivalent to the default behavior of
+     * runner-controlled sharding.
+     */
+    public Bound<T> withNumShards(int numShards) {
+      return new Bound<>(sink, Math.max(numShards, 0));
+    }
+
+    /**
+     * Writes all the elements in a bundle using a {@link Writer} produced by the
+     * {@link WriteOperation} associated with the {@link Sink}.
+     */
+    private class WriteBundles<WriteT> extends DoFn<T, WriteT> {
+      // Writer that will write the records in this bundle. Lazily
+      // initialized in processElement.
+      private Writer<T, WriteT> writer = null;
+      private final PCollectionView<WriteOperation<T, WriteT>> writeOperationView;
+
+      WriteBundles(PCollectionView<WriteOperation<T, WriteT>> writeOperationView) {
+        this.writeOperationView = writeOperationView;
+      }
+
+      @Override
+      public void processElement(ProcessContext c) throws Exception {
+        // Lazily initialize the Writer
+        if (writer == null) {
+          WriteOperation<T, WriteT> writeOperation = c.sideInput(writeOperationView);
+          LOG.info("Opening writer for write operation {}", writeOperation);
+          writer = writeOperation.createWriter(c.getPipelineOptions());
+          writer.open(UUID.randomUUID().toString());
+          LOG.debug("Done opening writer {} for operation {}", writer, writeOperationView);
+        }
+        try {
+          writer.write(c.element());
+        } catch (Exception e) {
+          // Discard write result and close the write.
+          try {
+            writer.close();
+            // The writer does not need to be reset, as this DoFn cannot be reused.
+          } catch (Throwable closeException) {
```

Comment:
Handle interrupted exception by checking if closeException is an InterruptedException and setting the interrupted bit on the current thread since you are suppressing the interrupt.
See http://www.ibm.com/developerworks/library/j-jtp05236/ for more details.